aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,202 @@
+
+    match ty_args[0] {
+        Type::U128 => {
+            // Get aggregator information and a value to subtract.
+            let value = safely_pop_arg!(args, u128);
+            let (id, limit) = aggregator_info_u128(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit)?;
+            Ok(smallvec![Value::bool(aggregator.try_sub(value).is_ok())])
+        },
+        Type::U64 => {
+            // Get aggregator information and a value to subtract.
+            let value = safely_pop_arg!(args, u64);
+            let (id, limit) = aggregator_info_u64(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit as u128)?;
+            Ok(smallvec![Value::bool(
+                aggregator.try_sub(value as u128).is_ok()
+            )])
+        },
+        _ => Err(PartialVMError::new(StatusCode::ABORTED)
+            .with_message("Unsupported type supplied to aggregator".to_string())
+            .with_sub_status(0x02_0005)
+            .into()),
+    }
+}
+
+/***************************************************************************************************
+ * native fun read<Element>(aggregator: &Aggregator<Element>): Element;
+ **************************************************************************************************/
+
+fn native_read(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert_eq!(args.len(), 1);
+
+    context.charge(AGGREGATOR_V2_READ_BASE)?;
+    let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
+    let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
+
+    match ty_args[0] {
+        Type::U128 => {
+            // Extract information from aggregator struct reference.
+            let (id, limit) = aggregator_info_u128(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit)?;
+            let value = aggregator.read_and_materialize(aggregator_context.resolver, &id)?;
+            Ok(smallvec![Value::u128(value)])
+        },
+        Type::U64 => {
+            let (id, limit) = aggregator_info_u64(&safely_pop_arg!(args, StructRef))?;
+            let aggregator = aggregator_data.get_aggregator(id, limit as u128)?;
+            let value = aggregator.read_and_materialize(aggregator_context.resolver, &id)?;
+            if value > u64::MAX as u128 {
+                return Err(PartialVMError::new(StatusCode::ABORTED)
+                    .with_message("Aggregator<u64>::read() output exceeds u64::MAX".to_string())
+                    .with_sub_status(0x02_0001)
+                    .into());
+            }
+            Ok(smallvec![Value::u64(value as u64)])
+        },
+        _ => Err(PartialVMError::new(StatusCode::ABORTED)
```

Comment:
This last case is repeated a lot :) Can we avoid that?

Reply:
With a match statement on an enum, you need to cover all the variants. The value can take several types. So, we do need an `_ => {something}` code.